pyo3-ffi/src/methodobject.rs

```diff
@@ -71,11 +78,44 @@ extern "C" {
 #[derive(Copy, Clone)]
 pub struct PyMethodDef {
     pub ml_name: *const c_char,
-    pub ml_meth: Option<PyCFunction>,
+    pub ml_meth: MlMeth,
     pub ml_flags: c_int,
     pub ml_doc: *const c_char,
 }
 
+/// Function types used to implement Python callables.
+///
+/// This union must be accompanied by the correct [ml_flags](PyMethodDef::ml_flags),
+/// otherwise the behavior is undefined.
+///
+/// See the [Python C API documentation][1] for more information.
+///
+/// [1]: https://docs.python.org/3/c-api/structures.html#implementing-functions-and-methods
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub union MlMeth {
+    /// This variant corresponds with [`METH_VARARGS`] *or* [`METH_NOARGS`] *or* [`METH_O`].
+    pub PyCFunction: Option<PyCFunction>,
```

Comment:
Can these things ever be null? Are the options necessary here?

Reply:
I don't think they can, a method slot always needs an implementation.

Reply:
I think so too. Note that strictly speaking C function pointers are nullable whereas Rust ones aren't, and `pyo3-ffi` handles this slightly weirdly by making the typedefs non-nullable and then wrapping them in `Option` in every struct /  function where they're used.

Here I think it's probably ok to drop the `Option`.